lib/introspection.js

```diff
@@ -71,28 +69,120 @@ const parseSignature = (
   return signature;
 };
 
-const introspect = (
-  // Introspect interface
-  namespace // hash of interfaces
-  // Returns: hash of hash of record, { method, title, parameters }
-) => {
+const parseFunction = (fn, text) => {
+  const fnData = fn + '';
+  const fnIndex = text.indexOf(fnData);
+  const lines = text.slice(0, fnIndex)
+    .split('\n')
+    .map(line => line.trim());
+  const comments = [];
+
+  for (let i = lines.length - 2; i >= 0; i--) {
+    if (lines[i].indexOf('// ') !== 0) break;
+    comments.unshift(lines[i].slice(3));
+  }
+  return comments;
+};
+
+const parseComments = (lines) => {
+  const sig = {
+    title: lines[0] || '',
+    description: '',
+    parameters: [],
+    comments: []
+  };
+  let index = 1;
+
+  while (index < lines.length && !lines[index].startsWith(' ')) {
+    sig.description += '\n' + lines[index++];
+  }
+
+  while (index < lines.length && lines[index].startsWith(' ')) {
+    let arg = lines[index++].trimEnd();
+    const len = arg.length;
+    arg = arg.trimStart();
+    const offset = len - arg.length;
+    const [name, rest] = common.section(arg, ' - ');
+
+    if (rest !== '') {
+      let [type, comment] = common.section(rest, ', ');
+      if (!ALL_TYPES.find(s => type.startsWith(s))) {
+        comment = type;
+        type = '';
+      }
+      sig.parameters.push({ name, type, comment, offset });
+    } else {
+      const parameter = sig.parameters.pop();
+      parameter.comment += '\n' + arg;
+      sig.parameters.push(parameter);
+    }
+  }
+
+  for (let i = index; i < lines.length; i++) {
+    if (NAMED_LINES.find(s => lines[i].startsWith(s))) {
+      const [name, comment] = common.section(lines[i], ':');
+      sig.comments.push({ name, comment });
+    } else {
+      const obj = sig.comments.pop();
+      obj.comment += '\n' + lines[i];
+      sig.comments.push(obj);
+    }
+  }
+  return sig;
+};
+
+// Introspect interface
+//   namespace - hash of interfaces
+// Returns: hash of hash of record, { method, title, parameters }
+const introspectOld = (namespace, text) => {
```

Comment:
Why do we need old versions?

Reply:
@tshemsedinov that's related to https://github.com/metarhia/metaschema/pull/107#issuecomment-423511723. We still have usages of old introspect in `guard` function (it gets type from the comment and checks for it).
I'd suggest to change the contract of guard to pass the argument types as an argument instead of getting them from the code-comment.
Otherwise we would have to keep in memory all of the files with definitions in order to get the code-commments OR pass filename to this function (and then require + source parse it) OR pass source of the file to this function (as it is now). Though I prefer signature change as all of the other variants are more like hacks and not the actual implementation.